marconi-chain-index/src/Marconi/ChainIndex/Experimental/Indexers/MintTokenEvent.hs

```diff
@@ -256,23 +348,191 @@ mkMintTokenIndexer dbPath = do
+  where
+  query point (QueryByAssetId policyId assetNameM eventType) ix = do
+    -- Filter events based on 'QueryByAssetId' query
+    let queryByAssetIdPredicate = Core.EventsMatchingQuery $ \(MintTokenBlockEvents events) ->
+          let isEventType :: MintTokenEvent -> Bool
+              isEventType e = case eventType of
+                Nothing -> True
+                Just MintEventType -> e ^. mintTokenEventAsset . mintAssetQuantity > 0
+                Just BurnEventType -> e ^. mintTokenEventAsset . mintAssetQuantity < 0
+              isAssetId :: MintTokenEvent -> Bool
+              isAssetId e =
+                case assetNameM of
+                  Nothing ->
+                    e ^. mintTokenEventAsset . mintAssetPolicyId == policyId
+                  Just assetName ->
+                    e ^. mintTokenEventAsset . mintAssetPolicyId == policyId
+                      && e ^. mintTokenEventAsset . mintAssetAssetName == assetName
+           in fmap MintTokenBlockEvents $
+                NonEmpty.nonEmpty $
+                  NonEmpty.filter (\e -> isAssetId e && isEventType e) events
+    -- TODO Find a way to propagate the error.
+    -- Without the 'runExceptT', there will be a conflict with the current MonadError constraint.
+    -- Currently, I just ignore the error, but ideally, we would want to propagate it.
+    timedEventsE <- runExceptT $ Core.query point queryByAssetIdPredicate ix
+    case timedEventsE of
+      Left _ -> pure []
+      Right timedEvents -> do
+        let sortedTimedEvents =
+              fmap
+                ( over
+                    (Core.event . mintTokenEvents)
+                    ( NonEmpty.sortBy
+                        ( \a1 a2 ->
+                            compare
+                              (a1 ^. mintTokenEventLocation . mintTokenEventIndexInBlock)
+                              (a2 ^. mintTokenEventLocation . mintTokenEventIndexInBlock)
+                        )
+                    )
+                )
+                $ List.sort timedEvents
+        pure sortedTimedEvents
+
+instance
+  (MonadIO m, MonadError (Core.QueryError (QueryByAssetId MintTokenBlockEvents)) m)
+  => Core.Queryable m MintTokenBlockEvents (QueryByAssetId MintTokenBlockEvents) Core.SQLiteIndexer
+  where
+  query =
+    let mintTokenEventQueryByAssetId :: Maybe Text -> SQL.Query
+        mintTokenEventQueryByAssetId whereClauseM =
+          SQL.Query $
+            Text.intercalate " " $
+              catMaybes
+                [ Just
+                    "SELECT slotNo, blockHeaderHash, txIndexInBlock, txId, policyId, assetName, quantity, redeemerHash, redeemerData"
+                , Just "FROM minting_policy_events"
+                , Just "WHERE slotNo <= :slotNo"
+                , fmap (\e -> Text.append " AND " e) whereClauseM
+                , Just "ORDER BY slotNo ASC, txIndexInBlock ASC"
+                ]
+     in -- TODO How about combining the NamedParam function with the Query?
```

Comment:
I thought it would be easier to separate them, especially when we reuse a `NamedParam` in different places in the query. We can chat about it and see how it'd look like though.